x/ibc/testing/chain.go

```diff
@@ -0,0 +1,298 @@
+func (chain *TestChain) GetContext() sdk.Context {
+	return chain.App.BaseApp.NewContext(false, chain.CurrentHeader)
+}
+
+// CommitNBlocks commits n blocks to state and updates the block height by 1 for each commit.
+func (chain *TestChain) CommitNBlocks(n int) {
+	for i := 0; i < n; i++ {
+		chain.App.Commit()
+		nextBlock(chain)
+	}
+
+}
+
+// CommitBlockWithNewTimestamp commits the current block and starts the next block with the
+// provided timestamp.
+func (chain *TestChain) CommitBlockWithNewTimestamp(timestamp int64) {
+	chain.App.Commit()
+	nextBlock(chain)
+	chain.App.BeginBlock(abci.RequestBeginBlock{
+		Header: abci.Header{
+			Height: chain.CurrentHeader.Height,
+			Time:   time.Unix(timestamp, 0),
+		}},
+	)
+}
+
+// nextBlock sets the last header to the current header and increments the current header to be
+// at the next block height and time.
+//
+// CONTRACT: this function must only be called after app.Commit() occurs
+func nextBlock(chain *TestChain) {
+	// set the last header to the current header
+	chain.LastHeader = ibctmtypes.CreateTestHeader(
+		chain.CurrentHeader.ChainID,
+		chain.CurrentHeader.Height,
+		chain.CurrentHeader.Time,
+		chain.Vals, chain.Signers,
+	)
+
+	// increment the current header
+	chain.CurrentHeader = abci.Header{
+		Height: chain.CurrentHeader.Height + 1,
+		Time:   time.Now(),
+	}
+
+	chain.App.BeginBlock(abci.RequestBeginBlock{Header: chain.CurrentHeader})
+}
+
+// SendMsg delivers a transaction through the application. It updates the senders sequence
+// number and updates the TestChain's headers.
+func (chain *TestChain) SendMsg(msg sdk.Msg) {
+	_, _, err := simapp.SignCheckDeliver(
+		chain.t,
+		chain.App.Codec(),
+		chain.App.BaseApp,
+		chain.GetContext().BlockHeader(),
+		[]sdk.Msg{msg},
+		[]uint64{chain.senderAccount.GetAccountNumber()},
+		[]uint64{chain.senderAccount.GetSequence()},
+		true, true, chain.senderPrivKey,
+	)
+	require.NoError(chain.t, err)
+
+	// SignCheckDeliver calls app.Commit()
+	nextBlock(chain)
+
+	// increment sequence for successful transaction execution
+	chain.senderAccount.SetSequence(chain.senderAccount.GetSequence() + 1)
+}
+
+// CreateClient will construct and execute a 07-tendermint MsgCreateClient. A counterparty
+// client will be created on the (target) chain.
+func (chain *TestChain) CreateClient(counterparty *TestChain) {
+	// commit counterparty's state
+	counterparty.CommitNBlocks(1)
+	chain.CommitNBlocks(1)
+
+	// construct MsgCreateClient using counterparty
+	msg := ibctmtypes.NewMsgCreateClient(
+		counterparty.ClientID, counterparty.LastHeader,
+		lite.DefaultTrustLevel, TrustingPeriod, UnbondingPeriod, MaxClockDrift,
+		chain.senderAccount.GetAddress(),
+	)
+
+	chain.SendMsg(msg)
+}
+
+// UpdateClient will construct and execute a 07-tendermint MsgUpdateClient. The counterparty
+// client will be updated on the (target) chain.
+func (chain *TestChain) UpdateClient(counterparty *TestChain) {
```

Comment:
Huge improvements!! Awesome that we're doing this instead of the previous approach